### DISCUSSION TITLE
Add support for first_activity

### DISCUSSION THREAD

**Discussion by DandyLyons** [SwiftPackageIndex/SwiftPackageIndex-Server#3356] [Ideas]
Context: https://iosdev.space/@dandylyons/113053286481579484

Allow users to filter packages by the first commit date.

**Comment by daveverwer** [comment]
This is a great idea, and makes perfect sense as an addition to what we already have.

Is it something you'd be interested in contributing? If so, I'll promote it to an issue to get it tracked.

**Reply by DandyLyons** [reply]
Yes I'm interested in contributing.

**Comment by DandyLyons** [comment]
I have a design question. Does it make sense to use first "activity" or first "commit"? 

I think as a user, I would be more interested in first "activity" (but I'm not sure). I care less about when the first commit was made. I care about when the first release was shipped (which I think is more or less equivalent to "activity" but I'm not sure). 

Then again, the distinction might not be meaningful enough to matter, and it looks like "first_commit_date" was already added to the DB.

**Reply by finestructure** [reply]
The advantage of `first_commit_date` is, as you say, that it's already there 😄

Computing a first activity similar to what we do for last activity (IIRC it's across PRs, issues) would be quite a bit more involved. I'm pretty sure we'd need to run additional Github graphQL queries to get the "head" of those stats (we currently only fetch the last few, again IIRC).

The difference is also probably not all that meaningful . If we're worried about it, we could "manually" look at first commit vs first activity for a number of packages to get a feel for it. (By manually I mean run the graphQL queries, get the first activity and compare it to the first commit date.)

I'd say go with the easier option for now!

**Reply by daveverwer** [reply]
Yes, I think I'd also expect the first commit date as I'd expect it to be very unusual for there to be other GitHub activity like opening issues before at least an initial commit happens? Is that even possible?

Either way, for the purposes of the intent of this feature, I think first commit date works well. We're not trying to be accurate down to a day. The primary usage of this feature will be, as you mention, to get an approximate "every package created since 2020", or similar.

The only thing I'd add is that I think we should keep the syntax for the search as `first_activity` rather than call it `first_commit`. There is a potential to call it `first_commit` if we also want to add a `last_commit`, which would make sense, but honestly I think that's overkill.

Keeping this as `first_activity` and `last_activity` will likely provide all the flexibility people need for a very long time.

I'll convert this to an issue now and close this discussion. Let's continue the conversation on the issue.

**Comment by daveverwer** [comment]
Tracked as #3362
